File: server/services/brokerAdapters/dxtrade.ts
/**
 * DXTrade adapter — fetches closed positions via REST API.
 * Credentials: loginId = username, server = broker API base URL,
 * passwordEnc decrypts to { password: "..." }
 */
import type { RawBrokerTrade } from '../brokerSyncService';

async function getToken(serverUrl: string, username: string, password: string): Promise<string> {
  const res = await fetch(`${serverUrl}/auth/token`, {
    method:  'POST',
    headers: { 'Content-Type': 'application/json' },
    body:    JSON.stringify({ username, password }),
  });
  if (!res.ok) throw new Error(`DXTrade auth failed: ${res.status}`);
  const data = await res.json() as any;
  const token = data.token ?? data.accessToken ?? data.access_token;
  if (!token) throw new Error('DXTrade: no token in auth response');
  return token;
}

async function getAccountId(serverUrl: string, token: string): Promise<string> {
  const res = await fetch(`${serverUrl}/user/accounts`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`DXTrade accounts: ${res.status}`);
  const data = await res.json() as any;
  const list = Array.isArray(data) ? data : (data.accounts ?? []);
  const id = list[0]?.id ?? list[0]?.accountId;
  if (!id) throw new Error('DXTrade: no account found');
  return String(id);
}

export async function fetchDXTradeTrades(
  serverUrl: string,
  username:  string,
  password:  string,
  fromMs:    number,
  toMs:      number,
): Promise<RawBrokerTrade[]> {
  const token   = await getToken(serverUrl, username, password);
  const accId   = await getAccountId(serverUrl, token);
  const from    = new Date(fromMs).toISOString().slice(0, 19);
  const to      = new Date(toMs).toISOString().slice(0, 19);

  const res = await fetch(
    `${serverUrl}/user/accounts/${accId}/positions?status=CLOSED&from=${from}&to=${to}&limit=500`,
    { headers: { Authorization: `Bearer ${token}` } },
  );
  if (!res.ok) throw new Error(`DXTrade positions: ${res.status}`);
  const data = await res.json() as any;
  const rows: any[] = Array.isArray(data) ? data : (data.positions ?? data.items ?? []);

  return rows.map(p => ({
    externalId:  String(p.positionId ?? p.id),
    symbol:      String(p.instrument ?? p.symbol ?? p.code),
    direction:   (p.side ?? p.orderSide ?? '').toUpperCase() === 'BUY' ? 'Long' : 'Short',
    lots:        parseFloat(p.qty ?? p.volume ?? p.quantity ?? 0) / 100_000,
    openPrice:   parseFloat(p.openPrice ?? p.entryPrice ?? 0),
    closePrice:  parseFloat(p.closePrice ?? p.exitPrice ?? 0),
    openTime:    p.openTime  ? new Date(p.openTime).getTime()  : fromMs,
    closeTime:   p.closeTime ? new Date(p.closeTime).getTime() : toMs,
    profit:      parseFloat(p.pnl ?? p.profit ?? p.realizedPnl ?? 0),
    commission:  -Math.abs(parseFloat(p.commission ?? p.fee ?? 0)),
    rawData:     p,
  }));
}

File: copy_platform/executors/dxtrade.py
"""
DXTrade executor — places / closes market orders via REST API.
Credentials: { loginId (username), password, server (broker API base URL) }
"""
import asyncio, logging
import aiohttp
from dataclasses import dataclass

log = logging.getLogger("executor.dxtrade")


@dataclass
class ExecResult:
    ok:          bool
    external_id: str | None = None
    entry_price: float | None = None
    error:       str | None = None


class DXTradeExecutor:
    def __init__(self, creds: dict, account_type: str):
        self.username = creds.get("loginId", "")
        self.password = creds.get("password", "")
        self.server   = (creds.get("server") or "").rstrip("/")

    async def _auth(self, session: aiohttp.ClientSession) -> tuple[str, str]:
        async with session.post(f"{self.server}/auth/token",
                                json={"username": self.username, "password": self.password}) as r:
            if r.status != 200:
                raise RuntimeError(f"DXTrade auth: {r.status}")
            data  = await r.json()
            token = data.get("token") or data.get("accessToken") or data.get("access_token")
            if not token:
                raise RuntimeError("DXTrade: no token")
            return token

    async def _get_account_id(self, session: aiohttp.ClientSession, token: str) -> str:
        async with session.get(f"{self.server}/user/accounts",
                               headers={"Authorization": f"Bearer {token}"}) as r:
            data = await r.json()
            rows = data if isinstance(data, list) else data.get("accounts", [])
            if not rows:
                raise RuntimeError("DXTrade: no accounts")
            return str(rows[0].get("id") or rows[0].get("accountId"))

    async def open_position(self, symbol: str, action: str,
                            volume_lots: float, sl: float | None,
                            tp: float | None) -> ExecResult:
        try:
            async with aiohttp.ClientSession() as session:
                token  = await self._auth(session)
                acc_id = await self._get_account_id(session, token)
                body = {
                    "orderCode":  symbol,
                    "orderQty":   int(volume_lots * 100_000),
                    "orderSide":  action,    # BUY / SELL
                    "orderType":  "MARKET",
                }
                if sl: body["stopLoss"]   = sl
                if tp: body["takeProfit"] = tp
                async with session.post(
                    f"{self.server}/user/accounts/{acc_id}/orders",
                    json=body,
                    headers={"Authorization": f"Bearer {token}"},
                ) as r:
                    data = await r.json()
                    if r.status not in (200, 201):
                        raise RuntimeError(f"DXTrade order: {r.status} {data}")
                    pos_id = str(data.get("positionId") or data.get("orderId") or "")
                    return ExecResult(ok=True, external_id=pos_id,
                                      entry_price=float(data.get("price") or 0))
        except Exception as e:
            log.error(f"[DXTrade] open_position failed: {e}")
            return ExecResult(ok=False, error=str(e))

    async def close_position(self, position_id: int,
                             volume_lots: float) -> ExecResult:
        try:
            async with aiohttp.ClientSession() as session:
                token  = await self._auth(session)
                acc_id = await self._get_account_id(session, token)
                async with session.post(
                    f"{self.server}/user/accounts/{acc_id}/positions/{position_id}/close",
                    json={"qty": int(volume_lots * 100_000)},
                    headers={"Authorization": f"Bearer {token}"},
                ) as r:
                    data = await r.json()
                    if r.status not in (200, 201):
                        raise RuntimeError(f"DXTrade close: {r.status} {data}")
                    return ExecResult(ok=True, external_id=str(position_id),
                                      entry_price=float(data.get("closePrice") or 0))
        except Exception as e:
            log.error(f"[DXTrade] close_position failed: {e}")
            return ExecResult(ok=False, error=str(e))

    async def modify_position(self, position_id: int,
                              sl: float | None, tp: float | None) -> ExecResult:
        try:
            async with aiohttp.ClientSession() as session:
                token  = await self._auth(session)
                acc_id = await self._get_account_id(session, token)
                body: dict = {}
                if sl is not None: body["stopLoss"]   = sl
                if tp is not None: body["takeProfit"] = tp
                async with session.patch(
                    f"{self.server}/user/accounts/{acc_id}/positions/{position_id}",
                    json=body,
                    headers={"Authorization": f"Bearer {token}"},
                ) as r:
                    return ExecResult(ok=r.status in (200, 204), external_id=str(position_id))
        except Exception as e:
            log.error(f"[DXTrade] modify_position failed: {e}")
            return ExecResult(ok=False, error=str(e))
